File: Go/XServer/base_test.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XServer

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/frameworkease/Universal.X.gRPC/Go/internal/Proto/Greet"
	"github.com/stretchr/testify/assert"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type GreetServer struct {
	Base
	awaked  bool
	started bool
	stopped bool
}

func (s *GreetServer) Awake() bool {
	Serve[GreetFooService](&Greet.FooService_ServiceDesc)
	Serve[GreetBarService](&Greet.BarService_ServiceDesc)
	Intercept(FooInterceptor)
	Intercept(BarInterceptor)
	s.awaked = true
	return true
}

func (s *GreetServer) Start() {
	s.Base.Start()
	s.started = true
}

func (s *GreetServer) Stop(wait *sync.WaitGroup) {
	s.Base.Stop(wait)
	s.stopped = true
}

type GreetFooService struct {
	Greet.UnimplementedFooServiceServer
}

func (s *GreetFooService) SayHi(ctx context.Context, req *Greet.RequestData) (*Greet.ResponseData, error) {
	return &Greet.ResponseData{Message: "Hello " + req.Name + "!"}, nil
}

type GreetBarService struct {
	Greet.UnimplementedBarServiceServer
}

func (s *GreetBarService) SayHello(ctx context.Context, req *Greet.RequestData) (*Greet.ResponseData, error) {
	return &Greet.ResponseData{Message: "Hi " + req.Name + "!"}, nil
}

var invokedInterceptors []string

var FooInterceptor = func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
	invokedInterceptors = append(invokedInterceptors, "FooInterceptor")
	return handler(ctx, req)
}

var BarInterceptor = func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
	invokedInterceptors = append(invokedInterceptors, "BarInterceptor")
	return handler(ctx, req)
}

func TestLifecycle(t *testing.T) {
	// 准备数据
	Configure(50051, 50052)
	serveUrl := fmt.Sprintf("127.0.0.1:%d", configuredServePort)
	probeUrl := fmt.Sprintf("http://127.0.0.1:%d", configuredProbePort)
	server := &GreetServer{}

	// 唤醒服务
	server.Awake()
	assert.True(t, server.awaked, "Greet 服务应当被唤醒。")

	// 启动服务
	server.Start()
	assert.True(t, server.started, "Greet 服务应当被启动。")

	// 等待服务
	{
		ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
		defer cancel()
		client := &http.Client{Timeout: 2 * time.Second}
		for {
			select {
			case <-ctx.Done():
				t.Fatal("Greet 服务启动超时")
			default:
			}
			resp, err := client.Get(probeUrl + "/health")
			if err == nil && resp.StatusCode == http.StatusOK {
				resp.Body.Close()
				break
			}
			if resp != nil {
				resp.Body.Close()
			}
			time.Sleep(100 * time.Millisecond)
		}
	}

	// 验证 healthz 端点
	{
		client := &http.Client{Timeout: 5 * time.Second}
		healthzResponse, err := client.Get(probeUrl + "/healthz")
		assert.NoError(t, err, "healthz 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, healthzResponse.StatusCode, "healthz 端点应当返回 200 OK。")
		healthzResponse.Body.Close()
	}

	// 验证 metrics 端点
	{
		client := &http.Client{Timeout: 5 * time.Second}
		metricsResponse, err := client.Get(probeUrl + "/metrics")
		assert.NoError(t, err, "metrics 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, metricsResponse.StatusCode, "metrics 端点应当返回成功状态码。")
		metricsContent := make([]byte, 1024)
		n, _ := metricsResponse.Body.Read(metricsContent)
		metricsResponse.Body.Close()
		content := string(metricsContent[:n])
		assert.True(t, strings.Contains(content, "go_gc_duration_seconds"), "metrics 端点内容应当包含指标数据。")
	}

	// 验证 pprof 端点
	{
		client := &http.Client{Timeout: 5 * time.Second}
		pprofResponse, err := client.Get(probeUrl + "/pprof/")
		assert.NoError(t, err, "pprof 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "pprof 端点应当返回 200 OK。")
		pprofResponse.Body.Close()

		pprofResponse, err = client.Get(probeUrl + "/pprof/allocs")
		assert.NoError(t, err, "allocs 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "allocs 端点应当返回 200 OK。")
		pprofResponse.Body.Close()

		pprofResponse, err = client.Get(probeUrl + "/pprof/block")
		assert.NoError(t, err, "block 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "block 端点应当返回 200 OK。")
		pprofResponse.Body.Close()

		pprofResponse, err = client.Get(probeUrl + "/pprof/cmdline")
		assert.NoError(t, err, "cmdline 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "cmdline 端点应当返回 200 OK。")
		pprofResponse.Body.Close()

		pprofResponse, err = client.Get(probeUrl + "/pprof/goroutine")
		assert.NoError(t, err, "goroutine 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "goroutine 端点应当返回 200 OK。")
		pprofResponse.Body.Close()

		pprofResponse, err = client.Get(probeUrl + "/pprof/heap")
		assert.NoError(t, err, "heap 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "heap 端点应当返回 200 OK。")
		pprofResponse.Body.Close()

		pprofResponse, err = client.Get(probeUrl + "/pprof/mutex")
		assert.NoError(t, err, "mutex 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "mutex 端点应当返回 200 OK。")
		pprofResponse.Body.Close()

		pprofResponse, err = client.Get(probeUrl + "/pprof/profile?seconds=1")
		assert.NoError(t, err, "profile 端点应当可以访问。")
		if assert.NotNil(t, pprofResponse, "profile 响应不应当为 nil") {
			assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "profile 端点应当返回 200 OK。")
			pprofResponse.Body.Close()
		}

		pprofResponse, err = client.Get(probeUrl + "/pprof/symbol")
		assert.NoError(t, err, "symbol 端点应当可以访问。")
		if assert.NotNil(t, pprofResponse, "symbol 响应不应当为 nil") {
			assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "symbol 端点应当返回 200 OK。")
			pprofResponse.Body.Close()
		}

		pprofResponse, err = client.Get(probeUrl + "/pprof/threadcreate")
		assert.NoError(t, err, "threadcreate 端点应当可以访问。")
		assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "threadcreate 端点应当返回 200 OK。")
		pprofResponse.Body.Close()

		pprofResponse, err = client.Get(probeUrl + "/pprof/trace?seconds=1")
		assert.NoError(t, err, "trace 端点应当可以访问。")
		if assert.NotNil(t, pprofResponse, "trace 响应不应当为 nil") {
			assert.Equal(t, http.StatusOK, pprofResponse.StatusCode, "trace 端点应当返回 200 OK。")
			pprofResponse.Body.Close()
		}
	}

	// 验证 grpc 服务
	{
		conn, err := grpc.NewClient(serveUrl, grpc.WithTransportCredentials(insecure.NewCredentials()))
		assert.NoError(t, err, "应当能够创建 gRPC 连接。")
		defer conn.Close()

		{
			client := Greet.NewFooServiceClient(conn)
			invokedInterceptors = nil
			response, err := client.SayHi(context.Background(), &Greet.RequestData{Name: "Alice"})
			assert.NoError(t, err, "Foo 服务应当正常执行。")
			assert.Equal(t, "Hello Alice!", response.Message, "Foo 服务应当返回正确的消息。")
			assert.Equal(t, "FooInterceptor", invokedInterceptors[0], "Foo 拦截器应当被调用。")
			assert.Equal(t, "BarInterceptor", invokedInterceptors[1], "Bar 拦截器应当被调用。")
		}

		{
			client := Greet.NewBarServiceClient(conn)
			invokedInterceptors = nil
			response, err := client.SayHello(context.Background(), &Greet.RequestData{Name: "Bob"})
			assert.NoError(t, err, "Bar 服务应当正常执行。")
			assert.Equal(t, "Hi Bob!", response.Message, "Bar 服务应当返回正确的消息。")
			assert.Equal(t, "FooInterceptor", invokedInterceptors[0], "Foo 拦截器应当被调用。")
			assert.Equal(t, "BarInterceptor", invokedInterceptors[1], "Bar 拦截器应当被调用。")
		}
	}

	// 停止服务
	{
		wait := &sync.WaitGroup{}
		server.Stop(wait)
		wait.Wait()
		assert.True(t, server.stopped, "Greet 服务应当被停止。")
	}
}

File: CSharp/tests~/Assets/Tests/Runtime/Scripts/XServer/base_test.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

#if !UNITY_5_3_OR_NEWER
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using FrameworkEase.Universal.X.gRPC;
using FrameworkEase.Universal.X.gRPC.Tests.Proto.Greet;
using Grpc.Core;
using Grpc.Core.Interceptors;
using Grpc.Net.Client;
using NUnit.Framework;

public partial class TestXServer
{
    internal class GreetServer : XServer.Base
    {
        internal bool awaked;
        internal bool started;
        internal bool stopped;

        public override bool Awake()
        {
            XServer.Serve(typeof(GreetFooService));
            XServer.Serve(typeof(GreetBarService));
            XServer.Intercept(typeof(GreetFooInterceptor));
            XServer.Intercept(typeof(GreetBarInterceptor));
            awaked = true;
            return true;
        }

        public override void Start()
        {
            base.Start();
            started = true;
        }

        public override void Stop(CountdownEvent counter)
        {
            base.Stop(counter);
            stopped = true;
        }
    }

    internal class GreetFooService : FooService.FooServiceBase
    {
        public override Task<ResponseData> SayHi(RequestData request, ServerCallContext context)
        {
            return Task.FromResult(new ResponseData { Message = $"Hello {request.Name}!" });
        }
    }

    internal class GreetBarService : BarService.BarServiceBase
    {
        public override Task<ResponseData> SayHello(RequestData request, ServerCallContext context)
        {
            return Task.FromResult(new ResponseData { Message = $"Hi {request.Name}!" });
        }
    }

    internal static readonly List<Interceptor> InvokedInterceptors = new();

    internal class GreetFooInterceptor : Interceptor
    {
        public override Task<TResponse> UnaryServerHandler<TRequest, TResponse>(TRequest request, ServerCallContext context, UnaryServerMethod<TRequest, TResponse> continuation)
        {
            InvokedInterceptors.Add(this);
            return base.UnaryServerHandler(request, context, continuation);
        }
    }

    internal class GreetBarInterceptor : Interceptor
    {
        public override Task<TResponse> UnaryServerHandler<TRequest, TResponse>(TRequest request, ServerCallContext context, UnaryServerMethod<TRequest, TResponse> continuation)
        {
            InvokedInterceptors.Add(this);
            return base.UnaryServerHandler(request, context, continuation);
        }
    }

    [Test]
    public async Task Lifecycle()
    {
        #region 准备数据
        XServer.Configure(51051, 51052);
        var serveUrl = $"http://127.0.0.1:{XServer.Configuration.ServePort}";
        var probeUrl = $"http://127.0.0.1:{XServer.Configuration.ProbePort}";
        var server = new GreetServer();
        #endregion

        #region 唤醒服务
        {
            server.Awake();
            Assert.That(server.awaked, Is.True, "Greet 服务应当被唤醒。");
        }
        #endregion

        #region 启动服务
        {
            server.Start();
            Assert.That(server.started, Is.True, "Greet 服务应当被启动。");

            #region 等待服务
            {
                using var httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(2) };
                var cts = new CancellationTokenSource(TimeSpan.FromSeconds(20));
                while (!cts.IsCancellationRequested)
                {
                    try
                    {
                        var response = await httpClient.GetAsync($"{probeUrl}/health", cts.Token);
                        if (response.IsSuccessStatusCode) break;
                    }
                    catch { }
                    await Task.Delay(100, cts.Token).ConfigureAwait(false);
                }
            }
            #endregion

            #region 验证 healthz 端点
            {
                using var client = new HttpClient();
                var healthzResponse = await client.GetAsync($"{probeUrl}/healthz");
                Assert.That(healthzResponse.IsSuccessStatusCode, Is.True, "healthz 端点应当返回成功状态码。");
                Assert.That(healthzResponse.StatusCode, Is.EqualTo(System.Net.HttpStatusCode.OK), "healthz 端点应当返回 200 OK。");
            }
            #endregion

            #region 验证 metrics 端点
            {
                using var client = new HttpClient();
                var metricsResponse = await client.GetAsync($"{probeUrl}/metrics");
                Assert.That(metricsResponse.IsSuccessStatusCode, Is.True, "metrics 端点应当返回成功状态码。");
                var metricsContent = await metricsResponse.Content.ReadAsStringAsync();
                Assert.That(metricsContent, Does.Contain("dotnet_collection_count_total"), "metrics 端点内容应当包含 'dotnet_collection_count_total'。");
            }
            #endregion

            #region 验证 grpc 服务
            {
                using var channel = GrpcChannel.ForAddress(serveUrl);

                {
                    InvokedInterceptors.Clear();
                    var client = new FooService.FooServiceClient(channel);
                    var response = client.SayHi(new RequestData { Name = "Alice" });
                    Assert.That(response.Message, Is.EqualTo("Hello Alice!"), "Foo 服务应当返回正确的消息。");
                    Assert.That(InvokedInterceptors[0], Is.InstanceOf<GreetFooInterceptor>(), "Foo 拦截器应当被调用。");
                    Assert.That(InvokedInterceptors[1], Is.InstanceOf<GreetBarInterceptor>(), "Bar 拦截器应当被调用。");
                }

                {
                    InvokedInterceptors.Clear();
                    var client = new BarService.BarServiceClient(channel);
                    var response = client.SayHello(new RequestData { Name = "Bob" });
                    Assert.That(response.Message, Is.EqualTo("Hi Bob!"), "Bar 服务应当返回正确的消息。");
                    Assert.That(InvokedInterceptors[0], Is.InstanceOf<GreetFooInterceptor>(), "Foo 拦截器应当被调用。");
                    Assert.That(InvokedInterceptors[1], Is.InstanceOf<GreetBarInterceptor>(), "Bar 拦截器应当被调用。");
                }
            }
            #endregion
            #endregion

            #region 停止服务
            {
                var counter = new CountdownEvent(1);
                server.Stop(counter);
                counter.Signal();
                counter.Wait(TimeSpan.FromSeconds(5));
                Assert.That(server.stopped, Is.True, "Greet 服务应当被停止。");
            }
            #endregion
        }
    }
}
#endif
